Sources/PackageGraph/RepositoryPackageContainer.swift

```diff
@@ -222,6 +222,7 @@ public class RepositoryPackageContainer: PackageContainer, CustomStringConvertib
         productFilter: ProductFilter
     ) throws -> (Manifest, [Constraint]) {
         let manifest = try self.loadManifest(at: revision, version: version)
+        let productFilter = manifest.toolsVersion < ToolsVersion.v5_2 ? ProductFilter.everything : productFilter
```

Comment:
This is a bit of a blunt instrument, but it should yield correct results.

Reply:
This single line fixes the top secret errors that were reported?

I’m curious if the errors occur generally during fresh resolution, or if they only occur in the presence of a pins file?

I ask because the way I implemented the filtering, manifests < 5.2 asked for every dependency product from every package. Essentially that meant any package that might contain the intended dependency would be resolved. The intent was that if a 5.1 manifest depends on a package that later gets a minor version bump with a 5.2 manifest, that transitive 5.2 manifest knows which of its products were actually requested and can therefore meaningfully filter its own dependencies. What occurs to me know is that had a side effect of culling some dependencies in a way that is logically valid, but not stable compared to previous treatment of 5.2 manifests.

For example, consider the case of a 5.2 manifest which declares a slew of package dependencies, but its actual targets declare no dependencies whatsoever. Under a 5.1 toolchain, all those dependencies were resolved. Under the initial filtering algorithm, all those dependencies were ignored. Ignoring them is valid as far as the build process is concerned. However, because it produces different pins, it introduces an instability where old and new toolchains would each be insisting that the other’s pins are invalid.

Can you look into it to see if this seems to be the real problem, or at least if it is the deciding factor that channels execution into the problematic code branch. If so, then the correct solution is to change the handling of pre‐5.2 manifests to request absolutely everything absolutely all the time.

Such a fix would be done most cleanly [here](https://github.com/apple/swift-package-manager/blob/4b6798f997058aae167c27cb4e42e22c7522afd2/Sources/PackageModel/Manifest.swift#L193), right where it originates, by injecting something like this:

```swift
if toolsVersion < ToolsVersion.v5_2 {
  return dependencies.map { $0.filtered(by: .everything) }
}
```

From that simple change, everything else should cascade back into place such that manifests from 5.1 and earlier are always and everywhere treated exactly as they were by the 5.1 toolchain. It would mean some graph pruning opportunities would missed, but it would ensure pin stability. Those missed optimizations would only effect old manifests, which presumably will become less and less common over time anyway. (5.2 manifests and above would still be properly skipping everything that isn’t used.)

Reply:
The issue I have been seeing is that we can miss dependencies of 5.1 and earlier packages that are actually required, the scenario is the same package being required by multiple different packages transitively. I need to dig in a bit more, I put this up more as a straw man so that we have _some_ solution to reactivate target-based dependency resolution that we can land in 5.4.

Where I am at right now, I'm seeing that we don't download a few dependencies as part of resolution which leads us to not being able to include them later on [here](https://github.com/apple/swift-package-manager/blob/main/Sources/Workspace/Workspace.swift#L1338). This piece of code in `Workspace` is kind of subtle, because unless we have previously fetched a given package, we'll just fail and ignore it. I'll report back once I have more.

Reply:
> I need to dig in a bit more, I put this up more as a straw man so that we have some solution to reactivate target-based dependency resolution that we can land in 5.4.

Yes, that makes sense.

I still think moving the workaround to the `Manifest` type like I suggested would be more consistent. I don’t think `RepositoryPackageContainer` should alter the resolution strategy, as it won’t apply to other conformers of the `PackageContainer` protocol.

> the scenario is the same package being required by multiple different packages transitively

That sounds like there is another place in need of a similar fix to [this part](https://github.com/apple/swift-package-manager/pull/3006/files#diff-134223fe78a1e46f08f8c1e9d166408b7105f941cb2cc4e844ebceb87a8f4ec8) of #3006. If a `topologicalSort` or similar graph traversal uses a node definition of just the package—instead of the package–filter pair—then it would think it had already visited that node and neglect to look at its dependencies even if the filter is different the second time the node shows up. The result would be exactly the symptoms you describe.